Firestore/Source/Core/FSTView.mm

```diff
@@ -154,45 +168,44 @@ @interface FSTView ()
 
 @property(nonatomic, strong) FSTDocumentSet *documentSet;
 
-/** Documents included in the remote target. */
-@property(nonatomic, strong) FSTDocumentKeySet *syncedDocuments;
-
-/** Documents in the view but not in the remote target */
-@property(nonatomic, strong) FSTDocumentKeySet *limboDocuments;
+@end
 
-/** Document Keys that have local changes. */
-@property(nonatomic, strong) FSTDocumentKeySet *mutatedKeys;
+@implementation FSTView {
+  /** Documents included in the remote target. */
+  DocumentKeySet _syncedDocuments;
 
-@end
+  /** Documents in the view but not in the remote target */
+  DocumentKeySet _limboDocuments;
 
-@implementation FSTView
+  /** Document Keys that have local changes. */
+  DocumentKeySet _mutatedKeys;
+}
 
-- (instancetype)initWithQuery:(FSTQuery *)query
-              remoteDocuments:(nonnull FSTDocumentKeySet *)remoteDocuments {
+- (instancetype)initWithQuery:(FSTQuery *)query remoteDocuments:(DocumentKeySet)remoteDocuments {
   self = [super init];
   if (self) {
     _query = query;
     _documentSet = [FSTDocumentSet documentSetWithComparator:query.comparator];
-    _syncedDocuments = remoteDocuments;
-    _limboDocuments = [FSTDocumentKeySet keySet];
-    _mutatedKeys = [FSTDocumentKeySet keySet];
+    _syncedDocuments = std::move(remoteDocuments);
+    _limboDocuments = DocumentKeySet{};
+    _mutatedKeys = DocumentKeySet{};
   }
   return self;
 }
 
-- (FSTViewDocumentChanges *)computeChangesWithDocuments:(FSTMaybeDocumentDictionary *)docChanges {
+- (FSTViewDocumentChanges *)computeChangesWithDocuments:
+    (const MaybeDocumentDictionary &)docChanges {
   return [self computeChangesWithDocuments:docChanges previousChanges:nil];
 }
 
-- (FSTViewDocumentChanges *)computeChangesWithDocuments:(FSTMaybeDocumentDictionary *)docChanges
+- (FSTViewDocumentChanges *)computeChangesWithDocuments:(const MaybeDocumentDictionary &)docChanges
                                         previousChanges:
                                             (nullable FSTViewDocumentChanges *)previousChanges {
   FSTDocumentViewChangeSet *changeSet =
       previousChanges ? previousChanges.changeSet : [FSTDocumentViewChangeSet changeSet];
   FSTDocumentSet *oldDocumentSet = previousChanges ? previousChanges.documentSet : self.documentSet;
 
-  __block FSTDocumentKeySet *newMutatedKeys =
-      previousChanges ? previousChanges.mutatedKeys : self.mutatedKeys;
```

Comment:
It's probably correct, but just to make sure: did this line in Objective-C make a shallow copy or a deep copy? What I'm getting at is that if it was a shallow copy, then `previousChanges.mutatedKeys` was updated, whereas in new code it's unaffected. I presume it was a deep copy, but want to verify.

Reply:
This is exactly the case where previously we were able to compute the view update in O(changes + lg(documents)) but this is now O(changes + documents), and every update is two full copies of the key set (once to populate `newMutatedKeys` in `computeChangesWithDocuments` and then again in copying from the change to `_mutatedKeys` in `applyChangesToDocuments)`.

I'm afraid that we really should make these collection types use `immutable::SortedSet` so that these algorithms continue to work as before.